In [41]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
import numpy as np
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [42]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [43]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [44]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [45]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_data.columns

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [46]:
df_market_data_scale = StandardScaler().fit_transform(df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d', 'price_change_percentage_14d', 'price_change_percentage_30d', 'price_change_percentage_60d', 'price_change_percentage_200d', 'price_change_percentage_1y']])

In [47]:
# Create a DataFrame with the scaled data
df_market_scale_pd = pd.DataFrame(df_market_data_scale,
                                  columns = ['price_change_percentage_24h', 'price_change_percentage_7d', 'price_change_percentage_14d', 'price_change_percentage_30d', 'price_change_percentage_60d', 'price_change_percentage_200d', 'price_change_percentage_1y'])

# Copy the crypto names from the original data
df_market_scale_pd['Crypto_names'] = df_market_data.index
# Set the coinid column as index
df_market_scale_pd = df_market_scale_pd.set_index('Crypto_names')

# Display sample data
df_market_scale_pd.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
Crypto_names                                                               
bitcoin                             0.508529                    0.493193   
ethereum                            0.185446                    0.934445   
tether                              0.021774                   -0.706337   
ripple                             -0.040764                   -0.810928   
bitcoin-cash                        1.193036                    2.000959   
binancecoin                         0.891871                    1.327295   
chainlink                           0.011397                    2.572251   
cardano                             0.102530                    1.508001   
litecoin                            0.077497                    0.334297   
bitcoin-cash-sv                     0.448952                   -0.190684   

                 price_change_percentage_14d  price_change_percentage_30d  \
Crypto_names                                                                
bitcoin                             0.772200                     0.235460   
ethereum                            0.558692                    -0.054341   
tether                             -0.021680                    -0.061030   
ripple                              0.249458                    -0.050388   
bitcoin-cash                        1.760610                     0.545842   
binancecoin                         0.800214                    -0.057148   
chainlink                           1.101647                    -0.490495   
cardano                             0.648885                     0.328959   
litecoin                            0.858520                    -0.012646   
bitcoin-cash-sv                    -0.248043                     0.051634   

                 price_change_percentage_60d  price_change_percentage_200d  \
Crypto_names                                                                 
bitcoin                            -0.067495                     -0.355953   
ethereum                           -0.273483                     -0.115759   
tether                              0.008005                     -0.550247   
ripple                             -0.373164                     -0.458259   
bitcoin-cash                       -0.291203                     -0.499848   
binancecoin                         0.778653                     -0.188232   
chainlink                          -0.931954                      0.387759   
cardano                            -0.486349                      0.065080   
litecoin                           -0.366477                     -0.486266   
bitcoin-cash-sv                    -0.529666                     -0.532961   

                 price_change_percentage_1y  
Crypto_names                                 
bitcoin                           -0.251637  
ethereum                          -0.199352  
tether                            -0.282061  
ripple                            -0.295546  
bitcoin-cash                      -0.270317  
binancecoin                       -0.225533  
chainlink                         -0.018284  
cardano                           -0.155428  
litecoin                          -0.292351  
bitcoin-cash-sv                   -0.206029

---

### Find the Best Value for k Using the Original Data.

In [48]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [49]:
# Create an empty list to store the inertia values
inertia_list = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters = i, random_state = 0)
    k_model.fit(df_market_scale_pd)
    inertia_list.append(k_model.inertia_)

C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMe

In [50]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {'k_original': k, 'inertia': inertia_list}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)

In [88]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_line_chart = df_elbow.hvplot.line(
    x = 'k_original',
    y = 'inertia',
    title = 'Elbow Curve',
    xticks = k
)

elbow_line_chart

:Curve   [k_original]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [89]:
# Initialize the K-Means model using the best value for k
k_model = KMeans(n_clusters = 4, random_state = 0)

In [90]:
# Fit the K-Means model using the scaled data
k_model.fit(df_market_scale_pd)

C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=0)

In [91]:
# Predict the clusters to group the cryptocurrencies using the scaled data
kmeans_predictions = k_model.predict(df_market_scale_pd)

# Print the resulting array of cluster values.
kmeans_predictions

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0])

In [92]:
# Create a copy of the DataFrame
df_market_scale_pd_predictions = df_market_scale_pd.copy()

In [93]:
# Add a new column to the DataFrame with the predicted clusters
df_market_scale_pd_predictions['Prediction'] = kmeans_predictions

# Display sample data
df_market_scale_pd_predictions.head()

price_change_percentage_24h  price_change_percentage_7d  \
Crypto_names                                                            
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
Crypto_names                                                             
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
Crypto_names                                                              
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  Prediction  
Crypto_names                                          
bitcoin                        -0.251637           2  
ethereum                       -0.199352           2  
tether                         -0.282061           0  
ripple                         -0.295546           0  
bitcoin-cash                   -0.270317           2

In [94]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
market_scale_scatter = df_market_scale_pd_predictions.hvplot.scatter(
    x = 'price_change_percentage_24h',
    y = 'price_change_percentage_7d',
    by = 'Prediction',
    hover_cols = ['Crypto_names']
)

market_scale_scatter

:NdOverlay   [Prediction]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Crypto_names)

---

### Optimize Clusters with Principal Component Analysis.

In [95]:
# Create a PCA model instance and set `n_components=3`.
pca_model = PCA(n_components = 3)

In [96]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
market_pca = pca_model.fit_transform(df_market_data)
# View the first five rows of the DataFrame. 
market_pca[:5]

array([[-341.80096268,  -51.36677548,   12.52547089],
       [-249.42046633,   24.11754777,  -14.23146597],
       [-402.61472077, -118.71073742,   24.83839662],
       [-406.75243715,  -79.48728629,    1.56633057],
       [-382.42994789, -103.43195906,   16.75307273]])

In [97]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca_model.explained_variance_ratio_

# Total Explained Variance
tpv = np.sum(pca_model.explained_variance_ratio_)
tpv

0.9998139161298989

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.9998 OR 99.98%

In [98]:
# Create a new DataFrame with the PCA data.
market_pca_df = pd.DataFrame(market_pca, columns = ['PC1', 'PC2', 'PC3'])
# Creating a DataFrame with the PCA data

# Copy the crypto names from the original data
market_pca_df['Crypto_names'] = df_market_data.index

# Set the coinid column as index
market_pca_df = market_pca_df.set_index('Crypto_names')

# Display sample data
market_pca_df.head(10)

PC1         PC2        PC3
Crypto_names                                      
bitcoin         -341.800963  -51.366775  12.525471
ethereum        -249.420466   24.117548 -14.231466
tether          -402.614721 -118.710737  24.838397
ripple          -406.752437  -79.487286   1.566331
bitcoin-cash    -382.429948 -103.431959  16.753073
binancecoin     -289.125020   12.287170  34.163848
chainlink         28.151408  154.987995 -73.126506
cardano         -174.519832   80.243493 -30.392830
litecoin        -406.613342  -91.783029   5.016144
bitcoin-cash-sv -311.219887 -143.285351   6.083080

---

### Find the Best Value for k Using the PCA Data

In [99]:
# Create a list with the number of k-values from 1 to 11
k_pca_data = list(range(1,11))

In [100]:
# Create an empty list to store the inertia values
inertia_list2 = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k_pca_data:
    k_model = KMeans(n_clusters = i, random_state = 0)
    k_model.fit(market_pca_df)
    inertia_list2.append(k_model.inertia_)

C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMe

In [101]:
# Create a dictionary with the data to plot the Elbow curve
elbow_curve = {'k_pca' : k_pca_data, 'inertia': inertia_list2}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_curve)
df_elbow_pca.head()

k_pca       inertia
0      1  6.997052e+07
1      2  8.180192e+06
2      3  2.580721e+06
3      4  8.237471e+05
4      5  4.559213e+05

In [102]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_pca_line_chart = df_elbow_pca.hvplot.line(
    x = 'k_pca',
    y = 'inertia',
    title = 'Elbow Curve',
    xticks = k
)

elbow_pca_line_chart

:Curve   [k_pca]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 2


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** It differs as the best k value found using the original data was 4 and the best k value of the PCA data is 2.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [103]:
# Initialize the K-Means model using the best value for k
k_model2 = KMeans(n_clusters = 2, random_state = 0)

In [104]:
# Fit the K-Means model using the PCA data
k_model2.fit(market_pca_df)

C:\Users\kevin\anaconda3\envs\myenv\lib\site-packages\sklearn\cluster\_kmeans.py:1446: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=2, random_state=0)

In [105]:
# Predict the clusters to group the cryptocurrencies using the PCA data
pca_predict = k_model2.predict(market_pca_df)
# Print the resulting array of cluster values.
pca_predict

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0])

In [106]:
# Create a copy of the DataFrame with the PCA data
pca_predict_df = market_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
pca_predict_df['pca_prediction'] = pca_predict

# Display sample data
pca_predict_df.head()

PC1         PC2        PC3  pca_prediction
Crypto_names                                                   
bitcoin      -341.800963  -51.366775  12.525471               0
ethereum     -249.420466   24.117548 -14.231466               0
tether       -402.614721 -118.710737  24.838397               0
ripple       -406.752437  -79.487286   1.566331               0
bitcoin-cash -382.429948 -103.431959  16.753073               0

In [107]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pca_predict_scatter = pca_predict_df.hvplot.scatter(
    x = 'PC1',
    y = 'PC2',
    by = 'pca_prediction',
    hover_cols = ['Crypto_names']
)

pca_predict_scatter

:NdOverlay   [pca_prediction]
   :Scatter   [PC1]   (PC2,Crypto_names)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [114]:
#For some reason, when trying to put them next to each other using +, this chart comes out to a straight line that's 0.
elbow_line_chart

:Curve   [k_original]   (inertia)

In [115]:
# Composite plot to contrast the Elbow curves
elbow_pca_line_chart

:Curve   [k_pca]   (inertia)

In [111]:
# Composite plot to contrast the clusters
market_scale_scatter + pca_predict_scatter

:Layout
   .NdOverlay.I  :NdOverlay   [Prediction]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,Crypto_names)
   .NdOverlay.II :NdOverlay   [pca_prediction]
      :Scatter   [PC1]   (PC2,Crypto_names)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** The impact of using fewer features to cluster the data using K-Means is significant according to the Crypto Data. Although using fewer features was simpler, it was not as accurate as the k value is different as well as the number of clusters. The K Model provided a k value of 4 and 4 clusters while the PCA Model provided a k value of 2 with 2 clusters. The PCA model was more accurate as the K Model chart shows 4 clusters, but 2 of the clusters only contain 1 data point for each cluster and the clusters were not separate entirely, making it hard to interpret. The PCA model shows a much easier to read graph and is more accurate. Thus the use of fewer features was less accurate.